addons/mass_mailing/models/mailing.py

```diff
@@ -650,3 +650,26 @@ def _image_to_url(b64image: bytes):
             return lxml.html.tostring(root)
 
         return body_html
+
+    def _prepare_mail_body(self, body=None):
+        """Prepare the email body before sending.
+        Add the text preview at the beginning of the mail. The preview text is displayed
+        bellow the mail subject of most mail client (gmail, outlook...).
+        We add the character `&zwnj;` (zero-width non-joiner) to fill the end of the
+        preview in order to not have the beginning of the mail at the end of the preview
+        (don't work with simple space as the content is trimmed).
+        https://litmus.com/blog/the-ultimate-guide-to-preview-text-support
+        """
+        self.ensure_one()
+
+        body = body or self.body_html
+        preview = re.findall(r'\>([^<>]+)\<', body)
```

Comment:
Isn't it better to build the preview while we are below 500 characters ? Just thinking about the people preparing a 20 Mb mailing with inlined images and tons of text, just to keep 500 first characters; we should try to keep a light weight processing as we already do a lot of extra processing with html ...

Reply:
@tde-banana-odoo 

If I compute the preview based on the first 500 characters, we might have no text (if an image is embedded at the beginning e.g.)

Maybe I can use `finditer` instead of `findall` and break when I have enough character ? (so we just search the character we want and not everything)

```python
import re
import timeit
import requests


def _prepare_mail_body():
    preview = re.findall(r"\>([^<>]+)\<", html)
    preview = " ".join([p.strip() for p in preview if p.strip()])[:500]
    return preview


def _prepare_mail_body_finditer():
    preview_elements = re.finditer(r"\>([^<>]+)\<", html)

    preview = ""
    for element in preview_elements:
        preview += element.group(0).strip()
        if len(preview) > 500:
            break

    return preview


html = requests.get(
    "https://demo.borland.com/testsite/stadyn_largepagewithimages.html"
).text

print("HTML Len:", len(html))


print(
    timeit.timeit(setup="from __main__ import html", stmt=_prepare_mail_body, number=50)
)
print(
    timeit.timeit(
        setup="from __main__ import html", stmt=_prepare_mail_body_finditer, number=50
    )
)
```

Results
```
HTML Len: 239012
0.12457598399487324
0.0008555829990655184
```

We are a lot faster, but the code is less readable

What do you think ?

Reply:
Indeed I thought of using an iterable instead of checking the whole content then stripping. I prefer the optimized solution, it is understandable and the performance difference is really interesting.